### ISSUE TITLE
"panic: attept to copy freed scalar" during destruction

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#11661]
Migrated from [rt.perl.org#99660](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=99660) (status was 'resolved')

Searchable as RT99660$

**Comment by p5pRT** [comment]
### From @doy

#### Created by @doy



Attempting to access a hash entry where an object was just removed from during that object's destructor causes a panic&#8203;:


\#\!/usr/bin/env perl
use strict;
use warnings;
use Test&#8203;::More;

my $called;

\{
&nbsp;    my %cache;
&nbsp;    sub get    \{ $cache\{$\_\[0\]\}         \}
&nbsp;    sub store  \{ $cache\{$\_\[0\]\} = $\_\[1\] \}
&nbsp;    sub remove \{ delete $cache\{$\_\[0\]\}; return \}
\}

\{
&nbsp;    package Thing;
&nbsp;    sub new \{ my $class = shift; bless \{ name => $\_\[0\] \}\, $class \}
&nbsp;    sub name \{ shift\->\{name\} \}
&nbsp;    sub DESTROY \{
&nbsp;        my $self = shift;
&nbsp;        my $thing = :&#8203;:get\($self\->name\);
&nbsp;        $called = 1;
&nbsp;    \}
\}

store\(foo => Thing\->new\('foo'\)\);
remove\('foo'\);

is\($called\, 1\, "got through the DESTROY method"\);

done\_testing;

\_\_END\_\_
&nbsp;        \(in cleanup\) panic&#8203;: attempt to copy freed scalar d5f500 to d5f4e8 at test\.pl line 21\.
not ok 1 \- got through the DESTROY method
\#   Failed test 'got through the DESTROY method'
\#   at test\.pl line 29\.
\#          got&#8203;: undef
\#     expected&#8203;: '1'
1\.\.1
\# Looks like you failed 1 test of 1\.



Bisecting points to this commit as causing the issue&#8203;:


commit f50383f58716bc0faa50de094d47cad8ad3fcbdb
Author&#8203;: Ton Hospel \<me\-02@&#8203;ton\.iguana\.be>
Date&#8203;:   Thu May 19 17&#8203;:05&#8203;:16 2011 \-0700

&nbsp;    \[perl \#85026\] Deleting the current iterator in void context
&nbsp;    
&nbsp;    Looking at the delete code I see another strange thing\. If the delete
&nbsp;    of the current iterator is done with the G\_DISCARD flag\, the corres\-
&nbsp;    ponding value is not freed and survives until the lazy deleted entry
&nbsp;    gets removed on the next hash iteration\. This is easily demonstrated
&nbsp;    like this&#8203;:
&nbsp;    
&nbsp;    perl \-wle '
&nbsp;    sub DESTROY \{ print "DESTROY" \}
&nbsp;    %a=\(a=>bless\[\]\);
&nbsp;    each %a;
&nbsp;    delete $a\{a\};
&nbsp;    print "END"
&nbsp;    '
&nbsp;    
&nbsp;    This prints&#8203;:
&nbsp;    
&nbsp;    END
&nbsp;    DESTROY
&nbsp;    
&nbsp;    notice the difference with&#8203;:
&nbsp;    
&nbsp;    perl \-wle '
&nbsp;    sub DESTROY \{ print "DESTROY" \}
&nbsp;    %hash = \(a => bless\[\]\);
&nbsp;    each %hash;
&nbsp;    $dummy = delete $hash\{a\}; $dummy = 0;
&nbsp;    print "END"
&nbsp;    '
&nbsp;    
&nbsp;    This prints&#8203;:
&nbsp;    
&nbsp;    DESTROY
&nbsp;    END
&nbsp;    
&nbsp;    This is easily solved by always replacing the deleted entry value with
&nbsp;    &PL\_sv\_placeholder\. It actually simplifies the code a bit except for the
&nbsp;    fact that the mro\_method\_changed from free\_hash\_ent now has to be done
&nbsp;    in hv\_delete

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl 5.15.3:

Configured by doy at Tue Sep 20 14:03:06 CDT 2011.

Summary of my perl5 (revision 5 version 15 subversion 3) configuration:
   
  Platform:
    osname=linux, osvers=3.0-arch, archname=x86_64-linux
    uname='linux zaon 3.0-arch #1 smp preempt tue aug 30 08:53:25 cest 2011 x86_64 intel(r) core(tm) i5 cpu m 520 @ 2.40ghz genuineintel gnulinux '
    config_args='-de -Dprefix=/home/doy/perl5/perlbrew/perls/perl-5.15.3 -Dusedevel'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.6.1 20110819 (prerelease)', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib/../lib /usr/lib/../lib /lib /usr/lib
    libs=-lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lc -lgdbm_compat
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.14.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.14'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -L/usr/local/lib -fstack-protector'

Locally applied patches:
    


@INC for perl 5.15.3:
    /home/doy/perl5/local/
    /home/doy/perl5/local/lib/perl5/site_perl/5.10.0/i686-linux
    /home/doy/perl5/perlbrew/perls/perl-5.15.3/lib/site_perl/5.15.3/x86_64-linux
    /home/doy/perl5/perlbrew/perls/perl-5.15.3/lib/site_perl/5.15.3
    /home/doy/perl5/perlbrew/perls/perl-5.15.3/lib/5.15.3/x86_64-linux
    /home/doy/perl5/perlbrew/perls/perl-5.15.3/lib/5.15.3
    .


Environment for perl 5.15.3:
    HOME=/home/doy
    LANG=en_US.UTF-8
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/doy/perl5/perlbrew/bin:/home/doy/perl5/perlbrew/perls/perl-5.15.3/bin:/home/doy/.bin/marathon:/home/doy/.bin/nethack:/home/doy/.bin:/usr/local/sbin:/usr/local/bin:/usr/lib/ccache/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/sbin:/usr/sbin:/sbin:/usr/bin/site_perl:/usr/bin/core_perl
    PERL5LIB=/home/doy/perl5/local/:/home/doy/perl5/local/lib/perl5/site_perl/5.10.0/i686-linux
    PERLBREW_HOME=/home/doy/.perlbrew
    PERLBREW_PATH=/home/doy/perl5/perlbrew/bin:/home/doy/perl5/perlbrew/perls/perl-5.15.3/bin
    PERLBREW_PERL=perl-5.15.3
    PERLBREW_ROOT=/home/doy/perl5/perlbrew
    PERLBREW_VERSION=0.28
    PERL_BADLANG (unset)
    PERL_CPANM_OPT=-q --mirror file:///home/doy/perl5/minicpan/ --mirror http://mirrors.kernel.org/cpan/ --mirror http://search.cpan.org/CPAN --prompt
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @cpansprout

Fixed locally; running tests\.\.\.\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

Fixed with commit 7058221\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
